voxel_grid/prototype/sfinae_check.hpp

```diff
@@ -0,0 +1,112 @@
+// #include <iostream>
+#include <optional>
+// #include <string>
+#include <type_traits>
+#include <vector>
+
+template <typename, typename>
+class TransformFilter;
+
+namespace detail {
+template <typename, typename>
+struct IsGridStruct;
+
+template <typename C, typename PointT>
+struct IsGridStruct {
+ private:
+  template <typename T>
+  static constexpr auto hasSetUp(T*) ->
+      typename std::is_same<decltype(std::declval<T>().setUp(
+                                std::declval<TransformFilter<T, PointT>*>())),
+                            bool>::type;
+  template <typename>
+  static constexpr std::false_type hasSetUp(...);
+
+  template <typename T>
+  static constexpr auto hasAddPoint(T*) -> typename std::is_same<
+      decltype(std::declval<T>().addPointToGrid(std::declval<PointT&>())),
+      void>::type;
+  template <typename>
+  static constexpr std::false_type hasAddPoint(...);
+
+  template <typename T>
+  static constexpr auto hasFilter(T*) -> typename std::is_same<
+      decltype(std::declval<T>().filterGrid(
+          std::declval<decltype(std::declval<T>().begin())>())),
+      std::optional<PointT>>::type;
+  template <typename>
+  static constexpr std::false_type hasFilter(...);
+
+  using has_set_up = decltype(hasSetUp<C>(0));
+  using has_add_point = decltype(hasAddPoint<C>(0));
+  using has_filter = decltype(hasFilter<C>(0));
+
+ public:
+  static constexpr bool is_set_up_valid = has_set_up::value;
+  static constexpr bool is_add_point_valid = has_add_point::value;
+  static constexpr bool is_filter_valid = has_filter::value;
+
+  // static constexpr bool is_valid = has_add_point::value && has_set_up::value
+  // && has_filter::value; static constexpr std::string print(){
+  //     return "testing " + is_set_up_valid;
+  // }
+  // static constexpr std::string status = print();
+};
+}  // namespace detail
+
+struct PointXYZ {};
+
+template <typename GridStruct, typename PointT>
+class TransformFilter {
+  using is_grid_struct = detail::IsGridStruct<GridStruct, PointXYZ>;
+  static_assert(is_grid_struct::is_set_up_valid,
+                "Required member is not satisfied: bool "
+                "setUp(TransformFilter<MyGridStruct,PointT>*)");
+  static_assert(
+      is_grid_struct::is_add_point_valid,
+      "Required member is not satisfied: void addPointToGrid(PointT& pt)");
+  static_assert(is_grid_struct::is_filter_valid,
+                "Required member is not satisfied: std::optional<PointXYZ> "
+                "filterGrid(...)");
```

Comment:
As the result, three static_assert are declared for better understanding from the client side